slam-libraries/viam-orb-slam3/orbslam_server_v1.cc

```diff
@@ -22,9 +25,43 @@ using viam::common::v1::PoseInFrame;
 #define MAX_COLOR_VALUE 255
 const std::string strRGB = "/rgb";
 const std::string strDepth = "/depth";
+const auto HEADERTEMPLATE =
+    "VERSION .7\n"
+    "FIELDS x y z\n"
+    // NOTE: If a float is more than 4 bytes
+    // on a given platform
+    // this size will be inaccurate
+    "SIZE 4 4 4\n"
+    "TYPE F F F\n"
+    "COUNT 1 1 1\n"
+    "WIDTH %d\n"
+    "HEIGHT 1\n"
+    "VIEWPOINT 0 0 0 1 0 0 0\n"
+    "POINTS %d\n"
+    "DATA binary\n";
 
 namespace viam {
 
+/*
+applies the mapSize to the HEADERTEMPLATE
+returning the pcd header as a string
+*/
+std::string pcdHeader(int mapSize) {
+    return str(boost::format(HEADERTEMPLATE) % mapSize % mapSize);
+}
+
+/*
+casts the float f to a pointer of unsigned 8 bit bytes
+iterates throught all the 8 bit bytes of f
+writes each 8 bit bytes to buffer
+*/
+void writeFloatToBufferInBytes(std::string &buffer, float f) {
+    auto p = (const uint8_t *)(&f);
+    for (std::size_t i = 0; i < sizeof(float); ++i) {
```

Comment:
@tessavitabile could I trouble you to confirm this is the safest way to get each of the 8 bytes out of a 32 bit float & add each byte to a string? I use a `uint8_t` pointer as I'm under the impression `const char *` (the pointer type the old implementation cast to) may not always be 8 bits on all platforms, which would break the expectations of our implementation, but please correct me if there is an error in my thinking / a different way I should do this.

Reply:
I'm not sure this is necessary here. We only support a couple platforms, and we test those platforms. If we were to strive for platform independence, I would want to test on more platforms (and we would also have other things to fix, such as assuming the size of a float in the pcd header). For now, I think we shouldn't make parts of the code platform-independent, if it adds complexity.

Reply:
We discussed offline that the motivation is readability. I don't feel that I can speak to the safest way to perform this--I would be looking it up for myself.

One suggestion I have is to add a check that sizeof(float)==4 and fail fast if this is not the case. Otherwise, LGTM

Reply:
For reference, Cartographer serialization uses this function:
```
void WriteSizeAsLittleEndian(uint64 size, std::ostream* out) {
  for (int i = 0; i != 8; ++i) {
    out->put(size & 0xff);
    size >>= 8;
  }
}
```
just as a reference but as Tess already pointed out I think we can focus for now on a single platform and your function can be expanded upon once we deal with more platforms.

Reply:
Thanks for the review y'all. Just to get another opinion I asked eric & he said that this is pretty much the only way to get each byte out of a float in C++ (which is not super safe no matter what one does) & is the same in terms of safety to the original.

As a result I'll keep this as it is a win for readability.

He did bring up 2, longer term (maybe for us to chat about later @tessavitabile) concerns, not with this implementation but more with the [PCD spec](https://pointclouds.org/documentation/tutorials/pcd_file_format.html) itself) which says:
`in binary form, where the data is a complete memory copy of the pcl::PointCloud.points array/vector. On Linux systems, we use mmap/munmap operations for the fastest possible read/write access to the data.`

What that means is that data corruption will likely silently occur if either of the following are true:
1.  the encoder & decoder processors have different endianness (i.e. not little endian)
2. the encoder & decoder processors have different floating point representations (i.e. not ieee 754)

Currently all our code assumes little endian (as that is what amd & intel support) & ieee 754.

This isn't relevant to this change but we may want to consider providing additional metadata in the PCD response regarding the floating point spec to use to interpret the bytes & what endianness to use as well, by checking on the PCD encoder side for these behaviors on boot.